unittests/test_variables.py

```diff
@@ -0,0 +1,152 @@
+# ReFrame Project Developers. See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+
+import pytest
+
+
+import reframe as rfm
+from reframe.core.fields import Field
+
+
+@pytest.fixture
+def novars():
+    class NoVars(rfm.RegressionTest):
+        pass
+    yield NoVars
+
+
+@pytest.fixture
+def onevar(novars):
+    class OneVar(novars):
+        var('foo', int, value=10)
+    yield OneVar
+
+
+def test_custom_var(onevar):
+    assert not hasattr(onevar, 'foo')
+    inst = onevar()
+    assert hasattr(onevar, 'foo')
+    assert isinstance(onevar.foo, Field)
+    assert hasattr(inst, 'foo')
+    assert inst.foo == 10
+
+
+def test_instantiate_and_inherit(novars):
+    inst = novars()
+    with pytest.raises(ValueError):
+        class MyTest(novars):
+            '''Error from name clashing'''
+
+
+def test_redeclare_var_clash(novars):
+    with pytest.raises(ValueError):
+        class MyTest(novars):
+            var('name', str)
+
+
+def test_inheritance_clash(novars):
+    class MyMixin(rfm.RegressionMixin):
+        var('name', str)
+
+    with pytest.raises(ValueError):
+        class MyTest(novars, MyMixin):
+            '''Trigger error from inheritance clash.'''
+
+
+def test_var_space_clash():
+    class Spam(rfm.RegressionMixin):
+        var('v0', int, value=1)
+
+    class Ham(rfm.RegressionMixin):
+        var('v0', int, value=2)
+
+    with pytest.raises(ValueError):
+        class Eggs(Spam, Ham):
+            '''Trigger error from var name clashing.'''
+
+
+def test_double_declare():
+    with pytest.raises(ValueError):
+        class MyTest(rfm.RegressionTest):
+            var('v0', int, value=1)
+            var('v0', float, value=0.5)
+
+
+def test_double_action_on_var():
+    with pytest.raises(ValueError):
+        class MyTest(rfm.RegressionTest):
+            set_var('v0', 2)
+            var('v0', int, value=2)
+
+
+def test_namespace_clash(novars):
+    with pytest.raises(ValueError):
+        class MyTest(novars):
+            var('current_environ', str)
+
+
+def test_set_var(onevar):
+    class MyTest(onevar):
+        set_var('foo', 4)
+
+    inst = MyTest()
+    assert not hasattr(onevar, 'foo')
+    assert hasattr(MyTest, 'foo')
+    assert hasattr(inst, 'foo')
+    assert inst.foo == 4
+
+
+def test_var_type(onevar):
+    class MyTest(onevar):
+        set_var('foo', 'bananas')
+
+    with pytest.raises(TypeError):
+        inst = MyTest()
+
+
+def test_set_undef(novars):
+    with pytest.raises(ValueError):
+        class MyTest(novars):
+            set_var('foo', 4)
+
+
+def test_require_var(onevar):
+    class MyTest(onevar):
+        require_var('foo')
+
+        def __init__(self):
+            print(self.foo)
+
+    with pytest.raises(AttributeError):
+        inst = MyTest()
+
+
+def test_required_var_not_present(onevar):
+    class MyTest(onevar):
+        require_var('foo')
+
+        def __init__(self):
+            pass
+
+    try:
```

Comment:
I also don't understand how this should pass. Isn't `foo` required here?

Reply:
> You don't need the `try/except` here if this is expected to work. If it throws an exception, the test will fail in any case.

Removed the `try/except` block.

> I also don't understand how this should pass. Isn't `foo` required here?

`foo` is required, but it is not referenced in the entire test, so this is all good. If a variable is required, the exception is thrown by the descriptor itself, because otherwise a required variable could be set in a hook at some point and then the value is already in there when the required var's value is accessed.
The only way I see we could be a bit more strict with this is to check at the very end of the test if the variable has been set or not and throw the error there if needed. Though, I'm not sure what the point of that would be, because that means that the test made it all the way to the end without any issues.

Reply:
> If a variable is required, the exception is thrown by the descriptor itself

You mean that the descriptor is not defined, so you will get an `AttributeError` or something, right? Which brings us to the point of the counter-intuitive inheritance behaviour discussed above, because in order to get this behaviour here, we need to define the descriptor too late, at the object creation. Perhaps, your other solution with the `__set__` counter inside the descriptor, for which I was not too much in favor at the beginning (😬) might be a better solution, because it could solve both problems.

Although the pattern of instantiating and then inheriting is not something that is done in any test, it's really scary that this implementation does not allow it! Even if you accidentally do it, you would be really wondering what's wrong with my test! Just adding this in the `hellocheck.py`, is enough to block everything:

```python
t = rfm.RegressionTest()
```

And the error message doesn't help.. 

```
unittests/resources/checks/hellocheck.py:13: 'name' already defined in class 'HelloTest'
```

Reply:
Also playing a bit with the `require_var`, I'm further less in favour of getting an `AttributeError`. Look at this:

```python
@rfm.simple_test
class HelloTest(rfm.RegressionTest):
    require_var('modules')

    def __init__(self):
        self.x = self.modules
        ...
```

```
./bin/reframe -c unittests/resources/checks/hellocheck.py -l
```

```
./bin/reframe: skipping test due to errors: HelloTest: use `-v' for more information
  FILE: /Users/karakasv/Repositories/reframe/unittests/resources/checks/hellocheck.py:25
[List of matched checks]
- CompileOnlyHelloTest (found in '/Users/karakasv/Repositories/reframe/unittests/resources/checks/hellocheck.py')
Found 1 check(s)
```

If I do `-v` I get an `AttributeError` for `modules`, which is not the best. I would prefer something like `you must set the 'modules' attribute`.